ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -83,2575 +83,1531 @@
 
-	if (string == NULL || result == NULL) {
-		return NULL;
+	if (string->len == 0) {
+		return result;
 	}
-	mbfl_string_init(result);
-	result->encoding = string->encoding;
-	mbfl_memory_device_init(&pc.device, MIN(string->len, width), 0);
-
-	/* output code filter */
-	pc.decoder = mbfl_convert_filter_new(
-	    &mbfl_encoding_wchar,
-	    string->encoding,
-	    mbfl_memory_device_output, 0, &pc.device);
-	pc.decoder_backup = mbfl_convert_filter_new(
-	    &mbfl_encoding_wchar,
-	    string->encoding,
-	    mbfl_memory_device_output, 0, &pc.device);
-	/* wchar filter */
-	encoder = mbfl_convert_filter_new(
-	    string->encoding,
-	    &mbfl_encoding_wchar,
-	    collector_strimwidth, 0, &pc);
-	if (pc.decoder == NULL || pc.decoder_backup == NULL || encoder == NULL) {
-		mbfl_convert_filter_delete(encoder);
-		mbfl_convert_filter_delete(pc.decoder);
-		mbfl_convert_filter_delete(pc.decoder_backup);
-		return NULL;
+
+	/* Skip over `from` characters and set start pointer */
+	unsigned char *start = string->val, *e = start + string->len;
+	if (from > 0) {
+		start += char_offset_to_byte_offset(string->encoding, string->val, string->len, from);
 	}
-	mkwidth = 0;
-	if (marker) {
-		mkwidth = mbfl_strwidth(marker);
+
+	/* Did we skip right to the end of the input string? */
+	if (start == e) {
+		return result; /* if so, return empty string */
 	}
-	pc.from = from;
-	pc.width = width - mkwidth;
-	pc.outwidth = 0;
-	pc.outchar = 0;
-	pc.status = 0;
-	pc.endpos = 0;
-
-	/* feed data */
-	p = string->val;
-	n = string->len;
-	if (p != NULL) {
-		while (n > 0) {
-			n--;
-			if ((*encoder->filter_function)(*p++, encoder) < 0) {
-				break;
-			}
-		}
-		mbfl_convert_filter_flush(encoder);
-		if (pc.status != 0 && mkwidth > 0) {
-			pc.width += mkwidth;
-			if (n > 0) {
-				while (n > 0) {
-					if ((*encoder->filter_function)(*p++, encoder) < 0) {
-						break;
-					}
-					n--;
-				}
-				mbfl_convert_filter_flush(encoder);
-			} else if (pc.outwidth > pc.width) {
-				pc.status++;
-			}
-			if (pc.status != 1) {
-				pc.status = 10;
-				pc.device.pos = pc.endpos;
-				mbfl_convert_filter_copy(pc.decoder_backup, pc.decoder);
-				mbfl_convert_filter_reset(encoder, marker->encoding, &mbfl_encoding_wchar);
-				p = marker->val;
-				n = marker->len;
-				while (n > 0) {
-					if ((*encoder->filter_function)(*p++, encoder) < 0) {
-						break;
-					}
-					n--;
-				}
-				mbfl_convert_filter_flush(encoder);
+
+	/* Now check how far we can get into remainder of string while still leaving
+	 * enough space for the trim marker, in case it is necessary to add it */
+	size_t marker_width = (marker ? mbfl_strwidth(marker) : 0);
+
+	size_t width = 0;
+	mbfl_convert_filter *filter = mbfl_convert_filter_new(string->encoding,
+		&mbfl_encoding_wchar, filter_count_width, 0, &width);
+
+	unsigned char *p = start, *enough_space_for_marker = start;
+	while (p < e) {
+		mbfl_filter_feed_char(*p++, filter);
+		if (width + marker_width <= max_width) {
```

Comment:
Style nit: Early break
```suggestion
		if (width + marker_width > max_width) {
			break;
		}
```

Reply:
Updating that...